### ISSUE TITLE
tensorrt export

### ISSUE THREAD

**Issue by SebastianJanampa** [ultralytics/ultralytics#21029]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello,

I am trying to export yolo11-pose to tensorrt to benchamark the inference time. Since I couldnt find a way to benchamark the modedl using ultralytics library. I used the script provided in the D-FINE repo.

```
python 
"""
Copyright (c) 2024 The D-FINE Authors. All Rights Reserved.
"""

import tensorrt as trt
import pycuda.driver as cuda
from utils import TimeProfiler
import numpy as np
import os
import time
import torch

from collections import namedtuple, OrderedDict
import glob
import argparse
from dataset import Dataset
from tqdm import tqdm


def parse_args():
    parser = argparse.ArgumentParser(description='Argument Parser Example')
    parser.add_argument('--infer_dir',
                        type=str,
                        default='./data/COCO2017/val2017',
                        help="Directory for images to perform inference on.")
    parser.add_argument("--engine_dir",
                        type=str,
                        default='trt_engines',
                        help="Directory containing model engine files.")
    parser.add_argument('--busy',
                        action='store_true',
                        help="Flag to indicate that other processes may be running.")
    args = parser.parse_args()
    return args

class TRTInference(object):
    def __init__(self, engine_path, device='cuda', backend='torch', max_batch_size=32, verbose=False):
        self.engine_path = engine_path
        self.device = device
        self.backend = backend
        self.max_batch_size = max_batch_size

        self.logger = trt.Logger(trt.Logger.VERBOSE) if verbose else trt.Logger(trt.Logger.INFO)
        self.engine = self.load_engine(engine_path)
        self.context = self.engine.create_execution_context()
        self.bindings = self.get_bindings(self.engine, self.context, self.max_batch_size, self.device)
        self.bindings_addr = OrderedDict((n, v.ptr) for n, v in self.bindings.items())
        self.input_names = self.get_input_names()
        self.output_names = self.get_output_names()

        if self.backend == 'cuda':
            self.stream = cuda.Stream()
        self.time_profile = TimeProfiler()
        self.time_profile_dataset = TimeProfiler()
        self.yolo = 'yolo' in engine_path

    def init(self):
        self.dynamic = False

    def load_engine(self, path):
        trt.init_libnvinfer_plugins(self.logger, '')
        with open(path, 'rb') as f, trt.Runtime(self.logger) as runtime:
            return runtime.deserialize_cuda_engine(f.read())

    def get_input_names(self):
        names = []
        for _, name in enumerate(self.engine):
            if self.engine.get_tensor_mode(name) == trt.TensorIOMode.INPUT:
                names.append(name)
        return names

    def get_output_names(self):
        names = []
        for _, name in enumerate(self.engine):
            if self.engine.get_tensor_mode(name) == trt.TensorIOMode.OUTPUT:
                names.append(name)
        return names

    def get_bindings(self, engine, context, max_batch_size=32, device=None):
        Binding = namedtuple('Binding', ('name', 'dtype', 'shape', 'data', 'ptr'))
        bindings = OrderedDict()
        for i, name in enumerate(engine):
            shape = engine.get_tensor_shape(name)
            dtype = trt.nptype(engine.get_tensor_dtype(name))

            if shape[0] == -1:
                dynamic = True
                shape[0] = max_batch_size
                if engine.get_tensor_mode(name) == trt.TensorIOMode.INPUT:
                    context.set_input_shape(name, shape)

            if self.backend == 'cuda':
                if engine.get_tensor_mode(name) == trt.TensorIOMode.INPUT:
                    data = np.random.randn(*shape).astype(dtype)
                    ptr = cuda.mem_alloc(data.nbytes)
                    bindings[name] = Binding(name, dtype, shape, data, ptr)
                else:
                    data = cuda.pagelocked_empty(trt.volume(shape), dtype)
                    ptr = cuda.mem_alloc(data.nbytes)
                    bindings[name] = Binding(name, dtype, shape, data, ptr)
            else:
                data = torch.from_numpy(np.empty(shape, dtype=dtype)).to(device)
                bindings[name] = Binding(name, dtype, shape, data, data.data_ptr())
        return bindings

    def run_torch(self, blob):
        for n in self.input_names:
            if self.bindings[n].shape != blob[n].shape:
                self.context.set_input_shape(n, blob[n].shape)
                self.bindings[n] = self.bindings[n]._replace(shape=blob[n].shape)

        self.bindings_addr.update({n: blob[n].data_ptr() for n in self.input_names})
        self.context.execute_v2(list(self.bindings_addr.values()))
        outputs = {n: self.bindings[n].data for n in self.output_names}
        return outputs

    def async_run_cuda(self, blob):
        for n in self.input_names:
            cuda.memcpy_htod_async(self.bindings_addr[n], blob[n], self.stream)

        bindings_addr = [int(v) for _, v in self.bindings_addr.items()]
        self.context.execute_async_v2(bindings=bindings_addr, stream_handle=self.stream.handle)

        outputs = {}
        for n in self.output_names:
            cuda.memcpy_dtoh_async(self.bindings[n].data, self.bindings[n].ptr, self.stream)
            outputs[n] = self.bindings[n].data

        self.stream.synchronize()

        return outputs

    def __call__(self, blob):
        if self.backend == 'torch':
            return self.run_torch(blob)
        elif self.backend == 'cuda':
            return self.async_run_cuda(blob)

    def synchronize(self):
        if self.backend == 'torch' and torch.cuda.is_available():
            torch.cuda.synchronize()
        elif self.backend == 'cuda':
            self.stream.synchronize()

    def warmup(self, blob, n):
        for _ in range(n):
            _ = self(blob)

    def speed(self, blob, n, nonempty_process=False):
        times = []
        self.time_profile_dataset.reset()
        for i in tqdm(range(n), desc="Running Inference", unit="iteration"):
            self.time_profile.reset()
            with self.time_profile_dataset:
                img = blob[i]
                if img['images'] is not None:
                    img['image'] = img['input'] = img['images'].unsqueeze(0)
                else:
                    img['images'] = img['input'] = img['image'].unsqueeze(0)
            with self.time_profile:
                _ = self(img)
            times.append(self.time_profile.total)

        # end-to-end model only
        if not self.yolo:
            print('end-to-end')
            times = sorted(times)
            if len(times) > 100 and nonempty_process:
                times = times[:100]

        avg_time = sum(times) / len(times)  # Calculate the average of the remaining times
        return avg_time

def main():
    FLAGS = parse_args()
    dataset = Dataset(FLAGS.infer_dir)
    im = torch.ones(1, 3, 640, 640).cuda()
    blob = {
            'image': im,
            'images': im,
            'input': im,
            'im_shape': torch.tensor([640, 640]).to(im.device),
            'scale_factor': torch.tensor([1, 1]).to(im.device),
            'orig_target_sizes': torch.tensor([[640, 640]]).to(im.device),
        }

    engine_files = glob.glob(os.path.join(FLAGS.engine_dir, "*.engine"))
    results = []

    for engine_file in engine_files:
        print(f"Testing engine: {engine_file}")
        model = TRTInference(engine_file, max_batch_size=1, verbose=False)
        model.init()
        model.warmup(blob, 400)
        t = []
        for _ in range(1):
            t.append(model.speed(dataset, 1000, FLAGS.busy))
        avg_latency = 1000 * torch.tensor(t).mean()
        results.append((engine_file, avg_latency))
        print(f"Engine: {engine_file}, Latency: {avg_latency:.2f} ms")

        del model
        torch.cuda.empty_cache()
        time.sleep(1)

    sorted_results = sorted(results, key=lambda x: x[0])
    for engine_file, latency in sorted_results:
        print(f"Engine: {engine_file}, Latency: {latency:.2f} ms")

if __name__ == '__main__':
    main()
``` 

I know this script works because i tested models like d-fine and rt-detr. However, when i run the run it with the engine file create using ultralytic command
```shell
yolo export model=yolo11n-pose.pt format=engine nms=True
```

i got the following error
```
[06/11/2025-22:07:16] [TRT] [E] IRuntime::deserializeCudaEngine: Error Code 1: Serialization (Serialization assertion header.magicTag == rt::PlanMagicTag::kPLAN_MAGIC_TAG failed.Trying to load an engine created with incompatible serialization version (542 != 1953657958). Check that the engine was not created using safety runtime, same OS was used and version compatibility parameters were set accordingly and that it is a TRT engine file.)
```

However, when i compile the onnx file using the following command
```
trtexec --onnx="yolo11n-pose.onnx" --saveEngine="yolo11n-pose.engine" --fp16
```

i can successfully benchmark yolo11-pose; but when i run and inference test, i notice that the results are not correct

![Image](https://github.com/user-attachments/assets/94936f88-8be5-423f-a685-100333c391f6)

Is there anything I'm missing? I thought that the onnx engine already contained postprocessing steps as scaling the results. 
Also, why do i need to export the yolo11-pose using ultralytics library instead of using trtexec? Does ultralytics apply anything else that i am missing?
Thanks in advance

**Comment by UltralyticsAssistant** [comment]
👋 Hello @SebastianJanampa, thank you for your detailed question and for using Ultralytics 🚀! This is an automated response to help get you started—an Ultralytics engineer will review your issue and assist you further soon.

We recommend checking out the [Docs](https://docs.ultralytics.com/) for helpful guides and reviewing our [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. Many common questions—including export details and benchmarking—are covered there.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) if you haven't already. This helps us debug more efficiently.

If you have a custom training ❓ Question, please share as much context as possible (dataset samples, training logs, export commands) and ensure you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community wherever you prefer:
- Real-time chat? [Discord](https://discord.com/invite/ultralytics) 🎧
- In-depth help? [Discourse](https://community.ultralytics.com/)
- Community sharing? [Subreddit](https://reddit.com/r/Ultralytics)

## Upgrade

Make sure you are running the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO models can be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for your patience while we review your question! If you have additional details, logs, or a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/), please add them here to help us assist you faster.

**Comment by glenn-jocher** [comment]
Hello @SebastianJanampa! The issue you're encountering is due to Ultralytics embedding metadata at the beginning of TensorRT engine files, which makes them incompatible with generic TensorRT runtime loaders.

When Ultralytics exports to TensorRT, it prepends metadata to the engine file for compatibility with the AutoBackend class. Your third-party script expects a pure TensorRT engine file, causing the deserialization error.

For your use case, you have a few options:

1. **Use Ultralytics benchmark mode** - this is the recommended approach:
```python
from ultralytics.utils.benchmarks import benchmark
benchmark(model="yolo11n-pose.pt", data="coco8-pose.yaml", imgsz=640, device=0)
```

2. **Strip metadata from Ultralytics engine** - modify your script to skip the metadata header when loading the engine file, similar to how AutoBackend handles it.

3. **Use the exported ONNX with proper post-processing** - when using `trtexec`, you'll need to implement the same post-processing that Ultralytics applies, including coordinate denormalization and keypoint scaling.

The reason for different results with `trtexec` is that Ultralytics applies specific post-processing steps during inference that aren't included in the raw ONNX model output. The Ultralytics export includes optimizations and compatibility layers that ensure consistent results across different backends.

**Comment by Y-T-G** [comment]
> I thought that the onnx engine already contained postprocessing steps as scaling the results.

There's no scaling inside the ONNX model. Just NMS. Scaling has to be done by the user.

**Comment by Y-T-G** [comment]
> I am trying to export yolo11-pose to tensorrt to benchamark the inference time. Since I couldnt find a way to benchamark the modedl using ultralytics library. 

You can benchmark models in Ultralytics in different formats.

https://docs.ultralytics.com/modes/benchmark/#usage-examples

**Comment by SebastianJanampa** [comment]
@Y-T-G ,

Could you provide me with an example of how to do the scaling?
As shown in the image of the first comment, the points are already scaled, but they are shifted. I didnt to anything except this

```shell

    im_pil = Image.open(file_path).convert("RGB")
    w, h = im_pil.size
    orig_size = torch.tensor([w, h])[None].to(device)

    transforms = T.Compose(
        [
            T.Resize((640, 640)),
            T.ToTensor(),
        ]
    )
    im_data = transforms(im_pil)[None]
    annotator = annotators[annotator_type](deepcopy(im_pil))

    blob = {
        "images": im_data.to(device),
    }

    output = m(blob)
    output = output['output0'][0] # 300, (4 + num_classes + num_keypoints*3)
    boxes, labels, keypoints = torch.split(output, [4, 2, 17*3], -1)
    scores = labels[:, 0]
    keypoints = keypoints.unflatten(-1,  (-1, 3))
```
